janus_server/src/datastore.rs

```diff
@@ -0,0 +1,318 @@
+use prio::codec::{decode_u16_items, encode_u16_items, CodecError, Decode};
+use tokio_postgres::{error::SqlState, IsolationLevel, Row};
+
+use crate::message::{Extension, HpkeCiphertext, Nonce, Report, TaskId, Time};
+use std::{future::Future, io::Cursor, pin::Pin};
+use tokio::sync::Mutex;
+
+// TODO(brandon): connection pooling (currently all concurrent transactions in a single Datastore are serialized)
+// TODO(brandon): prepare DB statements
+
+/// Datastore represents a datastore for Janus, with support for transactional reads and writes.
+/// In practice, Datastore instances are currently backed by a PostgreSQL database.
+pub struct Datastore {
+    db: Mutex<tokio_postgres::Client>,
+}
+
+impl Datastore {
+    /// new creates a new Datastore using the given Client for backing storage. It is assumed that
+    /// the Client is connected to a database with a compatible version of the Janus database schema.
+    pub fn new(db: tokio_postgres::Client) -> Datastore {
+        Self { db: Mutex::new(db) }
+    }
+
+    /// run_tx runs a transaction, whose body is determined by the given function. The transaction
+    /// is committed if the body returns a successful value, and rolled back if the body returns an
+    /// error value.
+    ///
+    /// The datastore will automatically retry some failures (e.g. serialization failures) by
+    /// rolling back & retrying with a new transaction, so the given function should support being
+    /// called multiple times. Values read from the transaction should not be considered as
+    /// "finalized" until the transaction is committed, i.e. after `run_tx` is run to completion.
+    pub async fn run_tx<F, T>(&self, f: F) -> Result<T, Error>
+    where
+        for<'a> F: Fn(&'a Transaction) -> Pin<Box<dyn Future<Output = Result<T, Error>> + 'a>>,
+    {
+        loop {
+            let rslt = self.run_tx_once(&f).await;
+            if let Some(err) = rslt.as_ref().err() {
+                if err.is_serialization_failure() {
+                    continue;
+                }
+            }
+            return rslt;
+        }
+    }
+
+    async fn run_tx_once<F, T>(&self, f: &F) -> Result<T, Error>
+    where
+        for<'a> F: Fn(&'a Transaction) -> Pin<Box<dyn Future<Output = Result<T, Error>> + 'a>>,
+    {
+        // Open transaction.
+        let mut guard = self.db.lock().await;
+        let tx = Transaction {
+            tx: guard
+                .build_transaction()
+                .isolation_level(IsolationLevel::Serializable)
+                .start()
+                .await?,
+        };
+
+        // Run user-provided function with the transaction.
+        // TODO(brandon): can we box/pin the return value, rather than requiring `f` to return a pre-boxed/pinned future?
```

Comment:
I doubt that you can `Box::new` the returned value because then you'd have `F: Fn(&'a Transaction) -> dyn Future<...>`, and that can't work because the compiler needs to know the size of returned values.

I hesitate to comment on the usage of `Pin` until I have spent more time [contemplating it on the tree of woe](https://www.youtube.com/watch?v=p48daZLNYFs).